File: Python/array_based.py
import csv

class ArrayList:
    def __init__(self):
        self._data = []

    def append(self, item):
        self._data.append(item)

    def get(self, index):
        if 0 <= index < len(self._data):
            return self._data[index]
        raise IndexError("Index out of bounds")

    def set(self, index, item):
        if 0 <= index < len(self._data):
            self._data[index] = item
        else:
            raise IndexError("Index out of bounds")

    def __len__(self):
        return len(self._data)

    def to_list(self):
        return list(self._data)

    def from_list(self, data_list):
        self._data = list(data_list)

    def clear(self):
        self._data = []

    def save_to_csv(self, filename):
        with open(filename, 'w', newline='') as file:
            writer = csv.writer(file)
            writer.writerows([[x] for x in self._data])


class ArrayQueue(ArrayList):
    def enqueue(self, item):
        self.append(item)

    def dequeue(self):
        if not self._data:
            raise IndexError("Queue is empty")
        return self._data.pop(0)

    def peek(self):
        if not self._data:
            raise IndexError("Queue is empty")
        return self._data[0]

    def is_empty(self):
        return len(self._data) == 0

class ArrayStack(ArrayList):
    def push(self, item):
        self.append(item)

    def pop(self):
        if not self._data:
            raise IndexError("Stack is empty")
        return self._data.pop()

    def peek(self):
        if not self._data:
            raise IndexError("Stack is empty")
        return self._data[-1]

    def is_empty(self):
        return len(self._data) == 0

File: Js/array_based.js
const fs = require('fs');

class ArrayList {
  constructor() {
    this._data = [];
  }

  append(item) {
    this._data.push(item);
  }

  get(index) {
    if (index >= 0 && index < this._data.length) {
      return this._data[index];
    }
    throw new Error("Index out of bounds");
  }

  set(index, item) {
    if (index >= 0 && index < this._data.length) {
      this._data[index] = item;
    } else {
      throw new Error("Index out of bounds");
    }
  }

  length() {
    return this._data.length;
  }

  toList() {
    return [...this._data];
  }

  fromList(dataList) {
    this._data = [...dataList];
  }

  clear() {
    this._data = [];
  }

  saveToCSV(filename) {
    console.log(`Salvando dados em ${filename}`);
    const rows = this._data.map(x => [x].join(',')); // transforma cada item em linha CSV
    const content = rows.join('\n');
    fs.writeFileSync(filename, content);
  }
}

class ArrayQueue extends ArrayList {
  enqueue(item) {
    this.append(item);
  }

  dequeue() {
    if (this._data.length === 0) {
      throw new Error("Queue is empty");
    }
    return this._data.shift();
  }

  peek() {
    if (this._data.length === 0) {
      throw new Error("Queue is empty");
    }
    return this._data[0];
  }

  isEmpty() {
    return this._data.length === 0;
  }
}

class ArrayStack extends ArrayList {
  push(item) {
    this.append(item);
  }

  pop() {
    if (this._data.length === 0) {
      throw new Error("Stack is empty");
    }
    return this._data.pop();
  }

  peek() {
    if (this._data.length === 0) {
      throw new Error("Stack is empty");
    }
    return this._data[this._data.length - 1];
  }

  isEmpty() {
    return this._data.length === 0;
  }
}

module.exports = { ArrayList, ArrayQueue, ArrayStack };
